### ISSUE TITLE
Adding sparse `addmv` support on CPU - Mac OS - Apple Silicon M2

### ISSUE THREAD

**Issue by tvercaut** [pytorch/pytorch#96972]
### 🚀 The feature, motivation and pitch

As discussed in https://github.com/pytorch/pytorch/issues/77764#issuecomment-1472510169 it would be helpful to have a CPU fallback for [`torch.addmv`](https://pytorch.org/docs/stable/generated/torch.addmv.html?highlight=addmv#torch.addmv) on Mac OS when using sparse matrices.

### Alternatives

Run on google colab or similar

### Additional context

Steps to reproduce the issue on Mac OS:
```python
import torch
print(f"Running PyTorch version: {torch.__version__}")

dtype = torch.float32
device = torch.device("cpu")
#device = torch.device("mps")
print(f"Using device: {device}")

#mat = torch.randn((4,4), dtype=dtype, device=device)
mat = torch.randn((4,4), dtype=dtype, device=device).relu().to_sparse_csr()
mvec = torch.randn((4,), dtype=dtype, device=device)
avec = torch.randn((4,), dtype=dtype, device=device)

ovec = torch.addmv(avec, mat, mvec)
print(ovec)
```
leading to
```
UserWarning: Sparse CSR tensor support is in beta state. If you miss a functionality in the sparse tensor support, please submit a feature request to https://github.com/pytorch/pytorch/issues. (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/SparseCsrTensorImpl.cpp:56.)
  mat = torch.randn((4,4), dtype=dtype, device=device).relu().to_sparse_csr()
Traceback (most recent call last):
  File "/Users/tomvercauteren/dev/HSI/hvs_hsi_pytorch/examples/01_hsi_images/mpstest.py", line 14, in <module>
    ovec = torch.addmv(avec, mat, mvec)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Calling addmv on a sparse CPU tensor requires compiling PyTorch with MKL. Please use PyTorch built MKL support.
```

It may be that MKL can be compiled for Mac OS (and thus shipped in the default pytorch distribution for mac) or maybe an less optimmised alternative needs to be found (e.g. Eigen).

**Comment by leovinus2001** [comment]
Have you looked whether you can do it via the Apple Accelerate framework? There were extensions beyond dense BLAS for sparse matrices but long ago since I used it it. IMHO thinking about Intel MKL on Apple Silicon is a waste of time as Apple has a team for such CPU libraries. Just like with Metal/MPS/GPU. Might be easier to just ask them. https://developer.apple.com/documentation/accelerate/creating_sparse_matrices

**Comment by tvercaut** [comment]
Thanks @leovinus2001. As discussed by @cpuhrsch in https://github.com/pytorch/pytorch/issues/77764#issuecomment-1472427272, my understanding is that the PyTorch team currently doesn't have the bandwidth to support the integration of additional platform-specific libraries to support sparse tensors.

So yes, tapping into Apple Accelerate would be neat but given these constraints and given that MKL is already used throughout PyTorch, even if not optimal on ARM / Silicon Macs, it could make sense to try it. If MKL can be used on Mac, it would certainly help "quickly" bridge the current gap in functionality.

On a related note, I just encountered another such missing functionality:
```
RuntimeError: Calling triangular_solve on a sparse CPU tensor requires compiling PyTorch with MKL. Please use PyTorch built MKL support.
```

Also interestingly, I realised that a workaround for the missing `addmv` can be to use `torch.mm` (with unsqueeze/squeeze as needed).

**Comment by malfet** [comment]
Is there a generic performant sparse library?  For example, can ARM Compute Library be used here? See https://developer.arm.com/documentation/101004/2030/Sparse-Linear-Algebra/Example-of-SpMV-usage?lang=en

**Comment by cpuhrsch** [comment]
@malfet - Can we ship mkldnn on Mac? It sounds like this more of a packaging problem assuming mkldnn is supported on mac.

**Comment by tvercaut** [comment]
Maybe a point of helpful clarification: What does MKL exactly mean in the context of PyTorch?

If it's what is now refered to as [OneMKL](https://github.com/oneapi-src/oneMKL) then this doesn't seem to have an ARM64 backend unfortunately (yet?). This is also confirmed [here](https://community.intel.com/t5/Intel-oneAPI-Math-Kernel-Library/Apple-Arm-Chip-Support/m-p/1445569).

If it's what is now being referred to as [OneDNN](https://github.com/oneapi-src/oneDNN) (and previously [MKL-DNN or DNNL](https://oneapi-src.github.io/oneDNN/dev_guide_transition_to_dnnl.html)), then this seems to support ARM64.

**Comment by leovinus2001** [comment]
Thanks for the clarifications! 

@tvercaut 
> Also interestingly, I realised that a workaround for the missing addmv() can be to use torch.mm (with unsqueeze/squeeze as needed).

Firstly, in that context, to go for a quick fix, it seems that IF torch.addmm () works fine on MPS then that is a potential solution for now. Just reshape the vector and voila. Ps: I have mapped sgemv() to sgemm () like that in the past ;)

Secondly, regarding MKL and lack of resources to support a new backend, it is a challenge. Obviously a topic for another thread. The recent MPS support is a great example though of what is possible. 

However, I would say that using Accelerate/CPU would be a quality improvement for PyTorch on macOS in future. Also for iOS. When I built PyTorch from scratch and see 3 possible BLAS options on macOS but not the native Accelerate then it just makes me sad. Using one alternative BLAS on macOS for debugging makes sense but to get the most speed and performance, the Accelerate solutions are the best and most forward compatible.

Finally,  am happy to help where I can. If you like a simple macOS code prototypes or proof-of-concept of addmv() or gemv() for dense and sparse then I am happy to contribute along the lines what we discussed earlier for torch.mm() and GEMM. Here I made some prototypes for earlier issues
https://github.com/pytorch/pytorch/issues/81185#issuecomment-1219761089
and
https://github.com/pytorch/pytorch/issues/81185#issuecomment-1229571793

**Comment by leovinus2001** [comment]
> @malfet - Can we ship mkldnn on Mac? It sounds like this more of a packaging problem assuming mkldnn is supported on mac.

A word of caution please. Whether right or wrong, it seems to me that the sole purpose of using MKLDNN on macOS is to get "something going". That might seem appealing short term but long term it is counter productive. MKLDNN was built with Intel in mind, and Apple has moved on to Apple Silicon and more. 

In principle, the goal of PyTorch macOS support is to please the PyTorch users with best performance on macOS right? That is always the Apple user perspective anyway "the best for the user" and the recent MPS support and code based on Apple and community work seems a good example. IMHO it seems to me that MKLDNN is not part of that road. Similar like my earlier observation on the Accelerate Sparse Matrix support and native support. 

Happy to help where I can. A nice question for another thread would be something like "what MKLDNN calls would be necessary anyway on macOS?""

**Comment by malfet** [comment]
> @malfet - Can we ship mkldnn on Mac? It sounds like this more of a packaging problem assuming mkldnn is supported on mac.

@cpuhrsch yeah, we can, but I though you need MKL(which is exclusively Intel), rather than MKLDNN (which indeed has some ARM support)

**Comment by cpuhrsch** [comment]
@malfet - Yes, sorry, you're right. I meant MKL.

**Comment by malfet** [comment]
@cpuhrsch - MKL is developed by Intel and distributed in binary only form, so there isn't much one can do to make it work on M1. But integrating with Accelerate doesn't sound like such a bad idea, as we already doing it to BLAS/LAPACK on M1.

**Comment by cpuhrsch** [comment]
@malfet - Sure, but maybe it can still run M1 even if very inefficiently. That seems like potentially an easier fix to unblock baseline coverage than integrating with a new backend library. We'll need to spend some time internally discussing how we best go about M1 coverage for sparsity.

**Comment by tvercaut** [comment]
Thanks @malfet, @cpuhrsch. I didn't realise Apple Accelerate was already integrated in PyTorch. I quick search shows that it actually even is used in critical functions such as ReLU:
https://github.com/pytorch/pytorch/blob/38b687ed4de5d74423ef0d0a60a4aa007d0c4ec9/caffe2/operators/relu_op.cc#L19-L33

So given that OneMKL doesn't (yet?) provide a generic or Apple-specific backend, relying on Apple Accelerate for `addmv`, `triangular_solve` and the like on MAC cpus seems like a reasonable option to me.

Alternatively, one could also rely on Eigen for a more generic but probably less optimised approach.

Anyway, I will not be the person implementing and maintaing this, so not my call :)

**Comment by tvercaut** [comment]
If I understand the suggestion of using the x86 build, I must say that running a complete pytorch-based project through [Rosetta 2](https://support.apple.com/en-gb/HT211861) doesn't sound very appealing to me.

I however do agree that having a reference portable CPU-based implementation of the sparse operations currently supported through OneMKL would make a lot of sense. I guess this would mean prioritising porting [/mkl/SparseBlasImpl.cpp](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/mkl/SparseBlasImpl.cpp) to [Eigen](https://eigen.tuxfamily.org/dox/group__Sparse__chapter.html) before porting it to say [Apple Accelerate](https://developer.apple.com/documentation/accelerate/creating_sparse_matrices?language=objc) or [Arm Performance Libraries](https://developer.arm.com/documentation/101004/2210/Sparse-Linear-Algebra?lang=en).

On the GPU side, unfortunately, there doesn't seem to be any sparse matrix support in MPS but maybe this will eventually be possible through a more generic vulkan based backend?

**Comment by malfet** [comment]
@tvercaut Eigen used to have a very complicated licensing rules, but I see that later one is just MPL license, so indeed this is a good idea. Also, Accelerate supports only float and double tensors, while Eigen would probably work with complex ones as well. In the meantime, I've made a very simple `addmv` implementation in https://github.com/pytorch/pytorch/pull/97353

**Comment by leovinus2001** [comment]
> > Accelerate has BLAS implementations for single and double complex as well.
> 
> The issue discussed here relates to sparse matrix operations. These are not covered by BLAS / LAPACK. As far as the Accelerate documentation goes, only float and double are supported for sparse matrices: https://developer.apple.com/documentation/accelerate/creating_sparse_matrices?language=objc

Good point :) Indeed I forgot for a moment that we are discussing sparse implementations here and there the complex type is missing in Accelerate it seems. My apologies. I will see whether I can file an improvement request with Apple on that.

> > It might be obvious to you all but there is another advantge to using the Apple Accelerate framework instead of a 3rd party Eigen, OpenBlas, etc library.
> 
> Benchmarking these is complex but sure, I would expect Apple libs to perform better on Apple hardware than third-party ones. My take is that for a project like PyTorch, it makes sense to get a generic implementation that works mostly everywhere before focusing on getting the most performance out of a single platform. So to me (but I have no say in PyTorch governance or strategy), it's not "one or the other" but maybe just "one before the other"...

We are very much in agreement here.  For something like PyTorch, two setups and CPU/BLAS libraries makes sense. One cross-platform and known library is great for bootstrapping, reference cases and unit tests. It would even work with Linux-on-M1-Mac. And one that uses the most high performance stuff on the target platform like what has been done now with GPU/MPS and PyTorch.

> Also my understanding is that for dense ops, PyTorch can already benefit from the most efficient BLAS / LAPACK implementation on any given platform by linking with the corresponding library. There shouldn't be any need to tap into an Apple-specific API there. I would even guess that PyTorch is already using Apple Accelerate's BLAS / LAPACK in the default binary distribution:
> 
> ```
> otool -L .pyenv/versions/3.11.2/envs/my-virtual-env-3.11.2/lib/python3.11/site-packages/torch/lib/libtorch_cpu.dylib | grep Accelerate        
> 	/System/Library/Frameworks/Accelerate.framework/Versions/A/Accelerate (compatibility version 1.0.0, current version 4.0.0)
> ```

To be honest, I am not sure about that. For example, I am not sure that the cblas_dgemm() from Accelerate is used by PyTorch latest on any Apple macOS or iOS setup. 

The "otool" command you mention is a great check but it only says that Accelerate is used for "something". As we saw earlier in our discussion, the current PyTorch uses a reference to vDSP_vthres() in Accelerate for ReLU which could entirely explain your otool output.

I did investigate this very question on cblas_dgemm()-via-Accelerate a few years ago, probably with PyTorch 1.07 or similar. I also used "otool" and built PyTorch from scratch on macOS and grep'd the source for the relevant calls. At that time, I was unable to find any reference to Accelerate GEMM operations in use on macOS. This might have been changed later of course. Also, when building PyTorch from scratch, the build scripts allowed me to build on macOS with MKL and OpenBlas GEMM but the fleeting hint at the Accelerate GEMMs, I was unable to make it work. Seemed like that was incomplete. Hence I concluded that cblas_sgemm() et al via Accelerate was unsupported at the time in PyTorch.

Personally, I would be most happy if even the dense-only cblas_sgemm() and cblas_dgemm() from Accelerate/CPU are in use on Apple Silicon by PyTorch latest. Will have another look at that. Or if anyone know for sure, feel free to confirm, thanks.

Back to sparse add_mv() examples :)

**Comment by malfet** [comment]
@tvercaut so `addmv` should be available on Apple silicon now, will see what can be done about triangular_solve later.

**Comment by tvercaut** [comment]
Somehow out of topic for this issue focused on CPU support but anyway. For what it's worth, I tried implementing a workaround for the lack of sparse-dense multiplication on MPS and got as far as having a proof of concept running with `PYTORCH_ENABLE_MPS_FALLBACK=1`. The ability of the hack to run on MPS mostly hinges on `scatter_reduce` which is not (yet?) supported on MPS. If anyone is intersted see below.

<details>
  <summary>Code snippet</summary>

```python
import torch

print(f"Running PyTorch version: {torch.__version__}")

torchdevice = torch.device("cpu")
if torch.cuda.is_available():
    torchdevice = torch.device("cuda")
    print("Default GPU is " + torch.cuda.get_device_name(torch.device("cuda")))
elif hasattr(torch.backends, "mps") and torch.backends.mps.is_available():
    torchdevice = torch.device("mps")
    print("mps backend available")
print("Running on " + str(torchdevice))

import numpy as np
import typing

# Hacky workaround for lack of CSR matrices on mps
class SparseCsrHack:
    _crow_indices: torch.Tensor
    _col_indices: torch.Tensor
    _values: torch.Tensor
    _size: typing.Union[list, tuple, torch.Size]
    is_sparse = True
    layout = torch.sparse_csr

    def __init__(
        self,
        crow_indices: torch.Tensor,
        col_indices: torch.Tensor,
        values: torch.Tensor,
        size: typing.Union[list, tuple, torch.Size],
    ):
        self._crow_indices = crow_indices
        self._col_indices = col_indices
        self._values = values
        self._size = torch.Size(size)

    def crow_indices(self) -> torch.Tensor:
        return self._crow_indices

    def col_indices(self) -> torch.Tensor:
        return self._col_indices

    def values(self) -> torch.Tensor:
        return self._values

    def size(self) -> torch.Size:
        return self._size

# Hacky workaround for lack of COO matrices on mps
class SparseCooHack:
    _indices: torch.Tensor
    _values: torch.Tensor
    _size: typing.Union[list, tuple, torch.Size]
    is_sparse = True
    layout = torch.sparse_coo

    def __init__(self, indices: torch.Tensor, values: torch.Tensor, size: typing.Union[list, tuple, torch.Size]):
        self._indices = indices
        self._values = values
        self._size = torch.Size(size)

    def indices(self) -> torch.Tensor:
        return self._indices

    def values(self) -> torch.Tensor:
        return self._values

    def size(self) -> torch.Size:
        return self._size


# Hacky and incomplete workaround for lack of spmv / spmm on mps
def spmv_hack(A, x):
    if not A.is_sparse:
        return A @ x
    if A.layout == torch.sparse_coo:
        A_row_idx, A_col_idx = A.indices()
    elif A.layout == torch.sparse_csr:
        A_col_idx = A.col_indices()
        A_crow_idx = A.crow_indices()
        # Uncompress row indices:
        A_row_idx = torch.repeat_interleave(A_crow_idx[1:] - A_crow_idx[:-1])
        # workaround for "UserWarning: MPS: no support for int64 repeats mask, casting it to int32"
        if A_row_idx.dtype != A_crow_idx.dtype:
            A_row_idx = A_row_idx.to(dtype=A_crow_idx.dtype)
    rep_x = torch.index_select(x, dim=0, index=A_col_idx)
    Arepx = A.values()[..., None] * rep_x
    etmp = torch.empty((A.size()[0], x.size()[1]), device=x.device, dtype=x.dtype)
    # PYTORCH_ENABLE_MPS_FALLBACK=1 is needed to avoid
    # NotImplementedError: The operator 'aten::scatter_reduce.two_out'
    b = etmp.scatter_reduce(0, A_row_idx[..., None].repeat(1, x.size()[1]), Arepx, reduce="sum", include_self=False)
    return b


# Super simple test
s1 = 10
s2 = 3
s3 = 2
A_dense = torch.randn((s1, s2), dtype=torch.float32, device=torchdevice).relu()
if A_dense.device.type != "mps":
    A_csr = A_dense.to_sparse_csr()
    A_coo = A_dense.to_sparse_coo()
else:
    A_csr_cpu = A_dense.cpu().to_sparse_csr()
    A_csr = SparseCsrHack(
        crow_indices=A_csr_cpu.crow_indices().to(A_dense.device),
        col_indices=A_csr_cpu.col_indices().to(A_dense.device),
        values=A_csr_cpu.values().to(A_dense.device),
        size=A_csr_cpu.size(),
    )
    A_coo_cpu = A_dense.cpu().to_sparse_coo()
    A_coo = SparseCooHack(
        indices=A_coo_cpu.indices().to(A_dense.device),
        values=A_coo_cpu.values().to(A_dense.device),
        size=A_coo_cpu.size(),
    )
x = torch.randn((s2, s3), dtype=torch.float32, device=torchdevice)

np.set_printoptions(precision=3)

b_ref = A_dense @ x
print(f"b_ref=\n{b_ref.T.cpu().numpy()}\n")

if x.device.type != "mps":
    b_fromcsr = A_csr @ x
    print(f"b_fromcsr=\n{b_fromcsr.T.cpu().numpy()}")
    print(f"allclose:{torch.allclose(b_fromcsr,b_ref)}\n")

b_fromcsrhack = spmv_hack(A_csr, x)
print(f"b_fromcsrhack=\n{b_fromcsrhack.T.cpu().numpy()}")
print(f"allclose:{torch.allclose(b_fromcsrhack,b_ref)}\n")

if x.device.type != "mps":
    b_fromcoo = A_coo @ x
    print(f"b_fromcoo=\n{b_fromcoo.T.cpu().numpy()}")
    print(f"allclose:{torch.allclose(b_fromcoo,b_ref)}\n")

b_fromcoohack = spmv_hack(A_coo, x)
print(f"b_fromcoohack=\n{b_fromcoohack.T.cpu().numpy()}")
print(f"allclose:{torch.allclose(b_fromcoohack,b_ref)}\n")

b_fromdensehack = spmv_hack(A_dense, x)
print(f"b_fromdensehack=\n{b_fromdensehack.T.cpu().numpy()}")
print(f"allclose:{torch.allclose(b_fromdensehack,b_ref)}\n")
```
</details>

Let me know if you think anything in there warrants creating a new issue. And apologies for the duplicate but I also mentioned this in https://github.com/pytorch/pytorch/issues/77764#issuecomment-1509824745

**Comment by tvercaut** [comment]
I have opened a new issue (#187389 ) for the missing `triangular_solve` functionality as this was not covered in the previous patches.
